### ISSUE TITLE
CouchPotato fails to download from magnet

### ISSUE THREAD

**Issue by murbaniak** [RuudBurger/CouchPotatoServer#5321]
Hi,
One of the most annoying problems with CouchPotato is it's inability to handle magnet links. There is more and more of these and there is no excuse to handle them incorrectly. What CouchPotato is trying to do is to download the torrent file from torcache.net from the magnet link. Quite often this does not work as not all torrents are cached. So that's it, game over.  No torrent, no download. But these days most torrent clients support downloading from magnet links. One solution is to pass a meta encoded uri to the torrent client. It works quite well. For black hole, all that is required is this:
        # log.error('Failed converting magnet url to torrent: %s', torrent_hash)
        # return False
        # ok, no torrent so we create a torrent file with magnet uri, let client handle it
        filedata = 'd10:magnet-uri%s:%se' % (len(magnet_link), magnet_link);
        return filedata
in magnetToTorrent function in core/_base/downloader/main.py
That's it, works like a charm with rtorrent.
I have no time to provide a proper pull request for this right now, so can someone provide a proper fix please ? Also, code have to be fixed for specific torrent client as to pass a magnet uri file via sgi for example.
Cheers,
Maciej

**Comment by CouchPotatoBot** [comment]
Thanks for the issue report! Before a real human comes by, please make sure your report has all the below criteria checked

- [ ] Fill in all the dots. They're there for a reason
- [ ] Post logs, either inline (for smaller logs) or using [Pastebin](http://pastebin.com/)
- [ ] Post full version information

Please make sure you also read [contribution guide](https://github.com/RuudBurger/CouchPotatoServer/blob/develop/contributing.md#issues) and followed all the steps. 
Make the title describe your issue. Having "CP not working" or "I get this bug" for 100 issues, isn't really helpful. My master will close issues if there isn't enough information. On a good day he will tag the issue on close with the reason (like `can't reproduce`), but usually he won't, the lazy asshat.

Sometimes my master seems like a grumpy cat and responds with short answers. This isn't (always) because he hates you, but because he's on mobile or busy fixing bugs. If something isn't clear, please let him know. Maybe he can teach me his awesome ways.

Thanks!

**Comment by RuudBurger** [comment]
Only when the downloader doesn't handle magnet links it tries to use torrentcache. Otherwise it will send the magnetlink to the downloader.
I didn't know that you can create a torrentfile with the magnet data inside it. Wondering if this is handled properly by all downloaders

**Comment by murbaniak** [comment]
Well, the only one I tested it (black hole) with is rtorrent 0.9.2. Works. But CouchPotato does not send magnet link to rtorrent, for sure. I tested it and it did not work. And rtorrent handles magnet links well. This code must be changed (core/downloaders/rtorrent_.py):
        # Try download magnet torrents
        if data.get('protocol') == 'torrent_magnet':
            filedata = self.magnetToTorrent(data.get('url'))
            if filedata is False:
                return False
I suspect you must pass magnet link to rtorrent in the same form as the black hole method:
'd10:magnet-uri%s:%se' % (len(magnet_link), magnet_link);
I will try to test it today

**Comment by murbaniak** [comment]
I was just thinking that to correctly do the black hole with magnets, the config option should be added, either to resolve the magnet to torrent or save magnet to torrent file and drop it in.

**Comment by RuudBurger** [comment]
Or maybe as a backup. Because when torcache works, why change it?

**Comment by murbaniak** [comment]
Yes, but I am not sure if this format of the file d10:magnet-uri... works with anything other than rtorrent. I googled around and found this format ever being mentioned in rtorrent related posts.

**Comment by D4nte** [comment]
Ah, please ignore my comments in the other issue... I'll try to do a Pull Request for that.

**Comment by nicolasnoble** [comment]
Created a pull request for this, referenced above.

**Comment by bignellrp** [comment]
Thanks for the suggestions here.   nicolasnoble's pull request was merged into the develop branch.  I backed up and deleted my old couchpotato folder (leaving the logs and settings folders alone) and replaced with the develop version using the following git command:

git clone -b develop git@github.com:CouchPotato/CouchPotatoServer.git

I had to update my /etc/default/couchpotato file so it included the word Server and i was ready to restart with the same init.d script.  Hopefully this helps someone with similar issues who might be a git noob like me.
